src/2017/roleassign.org

```diff
@@ -77,6 +78,30 @@
       return reqs
 #+END_SRC
 
+** But how does it choose the robot?
+- Actually Black Magic
+- A peek behind the curtain
+#+BEGIN_SRC python
+class RoleRequirements:
+    def __init__(self):
+        self.destination_shape = None      #Distance from destination
+        self.has_ball = False
+        self.chipper_preference_weight = 0 #importance of chipper
+        self.required_shell_id = None
+        self.previous_shell_id = None      #controls switching of robots
+        self.required = False              #Can't be dropped, causes assignment failure if too many
+        self.priority = 0                  #Higher is more important
+        self.require_kicking = False
+        self.robot_change_cost = 1.0
+#+END_SRC
+- Every behavior adds one of these
+
+- Constraints you add modify the cost of that behavior for a specific robot
+#+BEGIN_SRC python
+if req.destination_shape != None:
```

Comment:
`is not` rather than `!=`

Reply:
This is from our code lol, I didn't write it